xdem/volume.py

```diff
@@ -0,0 +1,49 @@
+from __future__ import annotations
+
+import matplotlib.pyplot as plt
+import numpy as np
+import pandas as pd
+
+
+def hypsometric_binning(ddem: np.ndarray, dem: np.ndarray):
+
+    assert ddem.shape == dem.shape
+    nan_mask = np.isnan(ddem) | np.isnan(dem)
+    ddem = ddem[~nan_mask]
+    dem = dem[~nan_mask]
+
+    old_dem = dem - ddem
+    elevation_range = np.percentile(old_dem, 99) - np.percentile(old_dem, 1)
+
+    bin_size = 50 if elevation_range > 500 else elevation_range / 10
+
+    bins = np.arange(old_dem.min(), old_dem.max() + bin_size, step=bin_size)
+    indices = np.digitize(old_dem, bins=bins)
+
+    medians = np.zeros(shape=bins.shape[0] - 1, dtype=ddem.dtype) + np.nan
+    means = medians.copy()
+    stds = medians.copy()
+    counts = np.zeros_like(medians, dtype=int)
+    for i in np.unique(indices):
+        values_in_bin = ddem[indices == i]
+        if values_in_bin.shape[0] == 0:
+            continue
+
+        medians[i - 1] = np.median(values_in_bin)
+        means[i - 1] = np.mean(values_in_bin)
+        stds[i - 1] = np.std(values_in_bin)
```

Comment:
Once more, it's all in old geoutils! ;-) You can just use any operator, and have a list of defaults you want to output:
https://github.com/GeoUtils/geoutils/blob/5c41b2a343d1f0fc161eed4426154656c06841a4/geoutils/geovector.py#L660

Reply:
Actually, the solution is a few blocks below. The line I gave is for initializing the arrays to Nan.

Reply:
I added an nmad column in 5853875. Would you both prefer to have it as a dynamic callable instead of all the columns right now?

Reply:
I think it's more flexible to use a callable yes. People might want the 25-75 percentile to plot the IQR for example etc. Then of course you can set the ones right now as default.

Reply:
I remade the function to take any callable in 80d9bae. Now it doesn't give as much per default, but the percentile could e.g. be given by passing `aggregation_function=lambda x: np.percentile(x, 99)`.
```bash
                                             value   count                                                                                                                                                                                 
(225.28848266601562, 275.2884826660156] -19.098885   724.0                                                                                                                                                                                 
(275.2884826660156, 325.2884826660156]  -36.254318  1034.0                                                                                                                                                                                 
(325.2884826660156, 375.2884826660156]  -31.608276   617.0                                                                                                                                                                                 
(375.2884826660156, 425.2884826660156]         NaN     NaN                                                                                                                                                                                 
(425.2884826660156, 475.2884826660156]  -19.853256  1294.0                                                                                                                                                                                 
(475.2884826660156, 525.2884826660156]  -15.140961  1321.0                                                                                                                                                                                 
(525.2884826660156, 575.2884826660156]  -10.407410   997.0                                                                                                                                                                                 
(575.2884826660156, 625.2884826660156]   -3.927429  1021.0                                                                                                                                                                                 
(625.2884826660156, 675.2884826660156]    2.108887   791.0                                                                                                                                                                                 
(675.2884826660156, 725.2884826660156]    1.071594   530.0                                                                                                                                                                                 
(725.2884826660156, 775.2884826660156]    0.579987   296.0                                                                                                                                                                                 
(775.2884826660156, 825.2884826660156]   -1.488831   302.0                                                                                                                                                                                 
(825.2884826660156, 875.2884826660156]    1.046814   282.0                                                                                                                                                                                 
(875.2884826660156, 925.2884826660156]    0.201599   248.0                                                                                                                                                                                 
(925.2884826660156, 975.2884826660156]   -2.662354   161.0
```
@rhugonnet , now the columns are gone! Should the columns still be there, but other aggregate functions could be added if specified? I would consider this most recent approach to be the "cleanest".

Reply:
> +1 I'm fine with having all the columns, this will never be a limiting factor in terms of computing

Not so confident as you Romain. For large sample, the median and NMAD could be quite slow to run...
Anyway, I would make at least mean or median + count the defaults.